GFDL generic_tracers with MOM6 NUOPC cap

dougiesquire opened the pull request:
This PR is a WIP to enable GFDL generic_tracers to be used with ACCESS-OM3. 

~The important changes in this PR are in the `MOM6` directory. Currently, edits to `CMEPS` and `CDEPS` are also included to add oxygen to the `jra55do` datm datamode, but these were mostly done to familiarize myself with the codebases and will likely be removed.~ UPDATE: I've now reverted my changes to CMEPS and CDEPS.

Details of the changes to MOM6 can be found in `MOM6/generic_tracer_nuopc.md`.

For a corresponding ACCESS-OM3 configuration, see [here](https://github.com/dougiesquire/MOM6-CICE6/tree/1deg_jra55do_ryf_bling). A few changes to Payu are required to run - see [here](https://github.com/dougiesquire/payu/tree/access-om3-bling).

Progress towards #42

micaeljtoliveira commented on MOM6/extra_sources/mom_cap_gtracer_flux.F90:
```diff
@@ -0,0 +1,649 @@
+
+  ! local variables
+  type(FmsNetcdfDomainFile_t), dimension(:), pointer :: ocn_bc_restart => NULL() !< Structures describing the restart files
+  integer                                            :: num_ocn_bc_restart = 0   !< The number of restart files to use
```
I'm not sure if this is what you intended here, but in Fortran initialization on declaration implies the `save` attribute. This means the value of this variable will be stored between calls to this function.

Here it probably does not matter, has I assume the value will be overwritten when calling `coupler_type_register_restarts`, but it's always good to be aware of this. So I would suggest to remove the initialization.

dougiesquire replied:
Thanks for the tip. I stole this from somewhere, so I'll check why it's being initialised there and remove if not necessary.

micaeljtoliveira replied:
If the code comes from somewhere else, just leave as it is. On the other hand, a comment about where the code comes from would be great.

micaeljtoliveira commented on MOM6/extra_sources/mom_cap_gtracer_flux.F90:
```diff
@@ -0,0 +1,649 @@
+
+  real :: ra,rl,tc
+
+  tc = tk-273.15
```
[Magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)) are evil ;)

dougiesquire replied:
Agreed, but I copied this subroutine from FMScoupler and have deliberately left things as unchanged as possible.

micaeljtoliveira replied:
No problem. It was not clear where the routine came from.

dougiesquire replied:
I'll add some more comments throughout

micaeljtoliveira reviewed (commented):
I've had a look at the whole set of changes. Overall all looks reasonable. In the process I noticed a few things and left some comments.

A more general question: are the changes to CDEPS and CMEPS generic enough that they could be accepted upstream?

dougiesquire commented:
Thanks for taking a look @micaeljtoliveira!

> A more general question: are the changes to CDEPS and CMEPS generic enough that they could be accepted upstream?

They're generic, but I doubt they'd be wanted. But as I mentioned in the PR description, I will probably remove these changes anyway. All they do is add oxygen to the `jra55do` datm datamode, mostly as a learning exercise for myself and as a test of the coupling I added. I doubt we want to keep this - it's probably easier/cleaner to prescribe oxygen using the data override stuff I added, rather than including a new stream in our datm.

dougiesquire commented on .gitmodules:
```diff
@@ -16,3 +16,6 @@
 	url = https://github.com/ESCOMP/WW3
+[submodule "MOM6/GFDL_generic_tracers"]
+	path = MOM6/GFDL_generic_tracers
+	url = https://github.com/COSIMA/GFDL_generic_tracers.git
```
Because I'm now adding WOMBAT to this repo. I don't know whether WOMBAT will be wanted upstream but I'll do the developments in our fork regardless.

micaeljtoliveira replied:
Okay, thanks for clarifying.

dougiesquire commented on MOM6/patches/mom_ocean_model_nuopc.F90.patch:
```diff
@@ -0,0 +1,116 @@
+   ! it also initializes statistical waves.
+   call MOM_wave_interface_init(OS%Time, OS%grid, OS%GV, OS%US, param_file, OS%Waves, OS%diag, OS%restart_CSp)
+ 
+-  if (associated(OS%grid%Domain%maskmap)) then
```
It's a little complicated, but in short, this and the [other changes made with this commit](https://github.com/COSIMA/access-om3/commit/f0abeb62a3f661695e46e40ff5e1a0191441b092) bring the domain handling in `mom_ocean_model_nuopc.F90` inline with the more recently updated FMS cap. The changes are needed in order to successfully register restarts for FMS coupler type.

dougiesquire commented on MOM6/patches/mom_cap.F90.patch:
```diff
@@ -1,6 +1,279 @@
+ public SetVM
+ 
+-!> Internal state type with pointers to three types defined by MOM.
++! > Internal state type with pointers to types defined by MOM.
```
Good catch. I'm curious what you think about having the derived type defined in the subsequent lines here change with the preprocessor flag. It feels yucky, but I'm not sure of a better approach.

micaeljtoliveira replied:
Yeah, not a huge fan, but I think it's acceptable. At least it makes it clear what things were added to support the generic tracers.

Looking again at the code, I would suggest putting the `ifdef` inside the type declaration:

```
type ocean_internalstate_type   
  type(ocean_public_type),       pointer :: ocean_public_type_ptr
  type(ocean_state_type),        pointer :: ocean_state_type_ptr
  type(ice_ocean_boundary_type), pointer :: ice_ocean_boundary_type_ptr
#ifdef _USE_GENERIC_TRACER
  type(coupler_2d_bc_type),      pointer :: coupler_2d_bc_type_ptr
+#endif
 end type
```

This reduces the number of duplicated lines and, in my opinion, makes it easier to understand the intent.

dougiesquire commented:
@angus-g, you mentioned you might be able to take a look at this. If you have time, any feedback would be much appreciated!

angus-g reviewed (commented):
All in all, this looks fairly good, in my opinion. Most of the heavy lifting seems to be done by existing code lifted from the FMS coupler, and just a bit of glue in the NUOPC layer to get it working.

My only real concern would be about keeping in sync with the FMS coupler for any bugfixes, if required? Having said that, this is probably still the best way of going about it, rather than pulling in the entire coupler as a dependency!

dougiesquire commented on MOM6/patches/mom_cap.F90.patch:
```diff
@@ -1,6 +1,272 @@
++
++  ! Restore ocean fields related to extra tracer fluxes from restart files
++  call get_MOM_input(dirs=dirs)
++  call gas_fields_restore(ocean_public%fields, ocean_public%domain, dirs%restart_input_dir)
```
Yup - this will only do anything if there is a restart file in the restart dir. Otherwise the fields will just retain their previously-initialised values.

dougiesquire commented on MOM6/patches/mom_cap.F90.patch:
```diff
@@ -1,6 +1,272 @@
+ 
+   endif
+ 
++  ! Set NUOPC attribute additional_restart_dir to RESTART/ if not defined
++  additional_restart_dir = "RESTART/"
++  call NUOPC_CompAttributeGet(gcomp, name="additional_restart_dir", value=cvalue, &
++        isPresent=isPresent, isSet=isSet, rc=rc)
++  if (ChkErr(rc,__LINE__,u_FILE_u)) return
++  if (isPresent .and. isSet) then
++    additional_restart_dir = slasher(cvalue)
++  else
++    call ESMF_LogWrite('MOM_cap:additional_restart_dir unset. Defaulting to '//trim(additional_restart_dir), &
++          ESMF_LOGMSG_INFO)
++  endif
++  call NUOPC_CompAttributeSet(gcomp, name="additional_restart_dir", value=additional_restart_dir, rc=rc)
++  if (chkerr(rc,__LINE__,u_FILE_u)) return
```
Yeah (although NUOPC only really specifies how the files are handled, MOM does the writing). The name is general because it's a general config parameter that in theory it could be used by any model component handled by NUOPC. I'll see what I can do to make this clearer. One option is just to hardcode the directory (it wouldn't be the first).

dougiesquire replied:
The main problem is that NUOPC wants the restart input and output directories to be `./` for all components. The restart filenames include a timestep to prevent files from being overwritten and a pointer file keeps track of the latest restart. The name of the additional restart file needed for the generic tracer fluxes is set within each generic_tracer module (i.e. I don't want to change it) and does not include a timestamp, so writing to and reading from MOMs ` restart_output_dir` and `restart_input_dir` (`=./`), respectively, causes problems.

dougiesquire commented on MOM6/patches/mom_cap_methods.F90.patch:
```diff
@@ -0,0 +1,98 @@
++
++        call ESMF_LogWrite(trim(subname)//': generic_tracer flux: '//trim(atm_fields%bc(n)%name)//&
++              ', coupling field '//trim(stdname), ESMF_LOGMSG_INFO)
++        allocate (work(isc:iec,jsc:jec))
```
Good point. Currently the cap is [only set up to handle one field](https://github.com/COSIMA/access-om3/blob/f345d91b21c5bc478b25262d3375622cb008813d/MOM6/extra_sources/mom_cap_gtracer_flux.F90#L214C1-L219C13) (co2 concentration for co2 flux), but there could be quite a few as we enable other generic tracers.

dougiesquire commented:
Thanks heaps for the review @angus-g. I share your concern about staying on top of changes to FMScoupler. But I'm not sure of a better approach. It makes me feel slightly better that the routines were ripping don't seem to change very often: they haven't changed at all since the files were renamed 3 years ago ([atmos_ocean_fluxes_calc.F90](https://github.com/NOAA-GFDL/FMScoupler/commits/main/full/atmos_ocean_fluxes_calc.F90), [atmos_ocean_dep_fluxes_calc.F90](https://github.com/NOAA-GFDL/FMScoupler/commits/main/full/atmos_ocean_dep_fluxes_calc.F90))

dougiesquire commented:
I've done some basic tests to show that these changes don't impact performance when no generic tracers are configured - see [here](https://github.com/COSIMA/access-om3/issues/155#issuecomment-2106705687).

@micaeljtoliveira I've made a couple of minor changes since your last review. Would you mind taking a quick look and approving if you're happy. I'd like to get this merged.

I'm not sure whether to squash this or not. There's useful info in many of the individual commits. But I made a bit of a mess of the history with the last few commits because the files being patched changed quite a bit and the conflicts were very difficult to resolve (I basically ended up regenerating the patches...).

dougiesquire commented:
@micaeljtoliveira after a few hours in rebase purgatory I've tidied up the commit history. The bad news is that you can no longer tell from this branch what's been added since your last review. The answer is not much. All that's really changed is that the `generic_WOMBATlite` tracer module is now included in the build (i.e. my most recent commit)

micaeljtoliveira commented on MOM6/extra_sources/mom_cap_gtracer_flux.F90:
```diff
@@ -0,0 +1,674 @@
+          call mpp_error(FATAL, trim(error_header) // ' Sizes of flux fields do not match')
+        endif
+
+        if (gas_fluxes%bc(n)%implementation .eq. 'ocmip2') then
```
I find this kind of multiple `if` statements easier to read if they are written as a `select case`:

```Fortran
select case (gas_fluxes%bc(n)%implementation)
case ('ocmip2')
...
case ('duce')
...
case ('johnson')
...
case default
...
end select
```

dougiesquire replied:
I agree, but this is code that I copied from FMScoupler and I'd like to keep it as close as possible to the original (in case the original is updated)

micaeljtoliveira commented on MOM6/patches/MOM_coupler_types.F90.patch:
```diff
@@ -0,0 +1,28 @@
+ use MOM_couplertype_infra, only : CT_set_data, CT_extract_data, CT_redistribute_data
+ use MOM_couplertype_infra, only : coupler_1d_bc_type, coupler_2d_bc_type, coupler_3d_bc_type
+-use MOM_couplertype_infra, only : ind_flux, ind_alpha, ind_csurf
++use MOM_couplertype_infra, only : ind_flux, ind_deltap, ind_kw, ind_flux0
```
Do I understand correctly that the only reason to have these use statements here is to make them available to other modules that use this one?

If so, I would think it's a better practice to have the use statements where they are actually needed. But I assume you are just following what seems to be common practice in MOM?

dougiesquire replied:
Yeah, this is a layer to avoid directly using routines directly from FMS. It enables FMS1 and FMS2 to be used, despite their different APIs.

micaeljtoliveira commented on MOM6/patches/mom_cap.F90.patch:
```diff
@@ -1,6 +1,271 @@
++
+   ocean_public%is_ocean_pe = .true.
++#ifdef _USE_GENERIC_TRACER
++  ! Initialise structures for extra tracer fluxes
```
Would there be a way of only calling the generic tracers code if needed? Probably the memory and CPU footprint of these calls is small, but still, would be nice to skip them if no generic tracers are needed.

dougiesquire replied:
Possibly, but I'll have to think about that. I haven't looked at the memory, but I did some [basic tests](https://github.com/COSIMA/access-om3/issues/166) that show performance isn't degraded with this code is included but no generic tracers are configured

micaeljtoliveira reviewed (commented):
@dougiesquire I had another look at this. Just a couple of minor comments. I think at this point you need someone who really knows the code very well if you want a more thorough review.

dougiesquire commented:
Thanks again @micaeljtoliveira. I completely agree that it would be great to have someone who really knows the code review this, but I've really struggled to find anyone. I've opened issues on both the NCAR/ and mom-ocean/MOM6 Github repos and had no response. Hopefully it will be easier to get eyes on this once we're an official MOM node, but what do you think about just getting this merged for now? I've been using this for a while now and it seems to be working as expected.

If you're happy to merge, can you please approve? It feels wrong to bypass the branch protections.

dougiesquire commented:
Snap! Thanks!
